src/server.ts

```diff
@@ -0,0 +1,183 @@
+/**
+ * HTTP server routing.
+ * By Mia.
+ * @author mia-pi-git
+ */
+import {Config} from './config-loader';
+import {Dispatcher, ActionError} from './dispatcher';
+import * as http from 'http';
+import * as https from 'https';
+
+/**
+ * API request output should not be valid JavaScript.
+ * This is to protect against a CSRF-like attack. Imagine you have an API:
+ *     https://example.com/getmysecrets.json
+ * Which returns:
+ *     {"yoursecrets": [1, 2, 3]}
+ *
+ * An attacker could trick a user into visiting a site overriding the
+ * Array or Object constructor, and then containing:
+ *     <script src="https://example.com/getmysecrets.json"></script>
+ *
+ * This could let them steal the secrets. In modern times, browsers
+ * are protected against this kind of attack, but our `]` adds some
+ * safety for older browsers.
+ *
+ * Adding `]` to the beginning makes sure that the output is a syntax
+ * error in JS, so treating it as a JS file will simply crash and fail.
+ */
+const DISPATCH_PREFIX = ']';
+
+export function toID(text: any): string {
+	if (text?.id) {
+		text = text.id;
+	} else if (text?.userid) {
+		text = text.userid;
+	}
+	if (typeof text !== 'string' && typeof text !== 'number') return '';
+	return ('' + text).toLowerCase().replace(/[^a-z0-9]+/g, '');
+}
+
+export class Router {
+	server: http.Server;
+	port: number;
+	awaitingEnd?: () => void;
+	closing?: Promise<void>;
+	activeRequests = 0;
+	constructor(port = (Config.port || 8000)) {
+		this.port = port;
+		const handle = (
+			req: http.IncomingMessage, res: http.ServerResponse
+		) => void this.handle(req, res);
+
+		this.server = Config.ssl
+			? https.createServer(Config.ssl, handle)
+			: http.createServer(handle);
```

Comment:
It feels like there should be both an HTTP server _and_ an HTTPS server, rather than only one.